ledger/persistedaccts_list.go

```diff
@@ -0,0 +1,127 @@
+package ledger
+
+// persistedAccountDataList represents a doubly linked list.
+// must initiate with newPersistedAccountList.
+type persistedAccountDataList struct {
+	root     persistedAccountDataListNode // sentinel list element, only &root, root.prev, and root.next are used
+	len      int                          // current list length excluding (this) sentinel element
+	freeList *persistedAccountDataList    // preallocated nodes location
+}
+
+type persistedAccountDataListNode struct {
+	// Next and previous pointers in the doubly-linked list of elements.
+	// To simplify the implementation, internally a list l is implemented
+	// as a ring, such that &l.root is both the next element of the last
+	// list element (l.Back()) and the previous element of the first list
+	// element (l.Front()).
+	next, prev *persistedAccountDataListNode
+
+	// The list to which this element belongs. (helps removing items fast)
+	list *persistedAccountDataList
+
+	Value *persistedAccountData
+}
+
+func newPersistedAccountList() *persistedAccountDataList {
+	l := new(persistedAccountDataList).init()
+	l.freeList = new(persistedAccountDataList).init()
+
+	return l
+}
+
+func (l *persistedAccountDataList) init() *persistedAccountDataList {
+	l.root.next = &l.root
+	l.root.prev = &l.root
+	l.len = 0
+	return l
+}
+
+func (l *persistedAccountDataList) allocateFreeNodes(numAllocs int) *persistedAccountDataList {
+	if l.freeList == nil {
+		return l
+	}
+	for i := 0; i < numAllocs; i++ {
+		l.freeList.pushFront(&persistedAccountData{})
```

Comment:
for all the entries in the freelist, it would be best to keep the Value as `nil`. ( when placed back on to the freelist, it should be reset to nil as well )